third_party/move/move-compiler-v2/src/pipeline/explicit_drop.rs

```diff
@@ -74,23 +81,128 @@ impl<'a> ExplicitDropTransformer<'a> {
+            Bytecode::Ret(..) | Bytecode::Abort(..) => self.emit_bytecode(bytecode.clone()),
             _ => {
-                let released_temps = self.released_temps_at(code_offset);
-                self.drop_temps(&released_temps, bytecode.get_attr_id())
+                self.emit_bytecode(bytecode.clone());
+                self.explicit_drops_at(code_offset, bytecode)
             },
         }
     }
 
+    fn transform_jump_at(&mut self, code_offset: CodeOffset, attr_id: AttrId, l: Label) {
+        if let Some((new_l, codes)) = self.process_jump(code_offset, attr_id, l) {
+            self.emit_bytecode(Bytecode::Jump(attr_id, new_l));
+            self.emit_bytecodes(codes.into_iter())
+        } else {
+            self.emit_bytecode(Bytecode::Jump(attr_id, l))
+        }
+    }
+
+    fn transform_branch_at(
+        &mut self,
+        code_offset: CodeOffset,
+        attr_id: AttrId,
+        l1: Label,
+        l2: Label,
+        cond: TempIndex,
+    ) {
+        match (
+            self.process_jump(code_offset, attr_id, l1),
+            self.process_jump(code_offset, attr_id, l2),
+        ) {
+            (None, Some((new_l2, codes2))) => {
+                self.emit_bytecode(Bytecode::Branch(attr_id, l1, new_l2, cond));
+                self.emit_bytecodes(codes2.into_iter())
+            },
+            (Some((new_l1, codes1)), None) => {
+                self.emit_bytecode(Bytecode::Branch(attr_id, new_l1, l2, cond));
+                self.emit_bytecodes(codes1.into_iter())
+            },
+            (Some((new_l1, codes1)), Some((new_l2, codes2))) => {
+                self.emit_bytecode(Bytecode::Branch(attr_id, new_l1, new_l2, cond));
+                self.emit_bytecodes(codes1.into_iter());
+                self.emit_bytecodes(codes2.into_iter())
+            },
+            (None, None) => self.emit_bytecode(Bytecode::Branch(attr_id, l1, l2, cond)),
+        }
+        // TODO: if we need to drop primitives one day, then we need to drop the locals released by the branch instruction as well
+    }
+
+    /// `pred_offset`: refers to an instruction which may jumps to `label`
+    /// Returns `None` if no drops are inferred between (after state of) `pred_offset` and (before state of) `label`
+    /// Otherwise returns
+    /// - a fresh label
+    /// - a new code block that
+    ///     - starts with the fresh label
+    ///     - drops inferred locals
+    ///     - jumps to `label`
+    pub fn process_jump(
+        &mut self,
+        pred_offset: CodeOffset,
+        attr_id: AttrId,
+        label: Label,
+    ) -> Option<(Label, Vec<Bytecode>)> {
+        let suc_offset = self.label_offsets.get(&label).expect("label offset");
+        let temps_to_drop = &self.released_temps_between(pred_offset, *suc_offset);
+        if temps_to_drop.is_empty() {
+            None
+        } else {
+            Some(self.explicit_drops_before_label(label, temps_to_drop, attr_id))
```

Comment:
It seems that it would be simpler to (1) assume that the CFG has no critical edges (edges from a node with multiple successors to a node with multiple predecessors), and (2) always place this code in the successor blocks if needed, rather than having to worry about adding more blocks during this pass.
Otherwise, we might wind up with similar kludgey code for multiple transformation passes.

We can implement a prepass to break critical edges, and a postpass to collapse empty blocks and jumps to jumps, and then all the other intervening passes can just assume the CFG is stable.  What do you think?